libs/wire-subsystems/src/Wire/MiniBackend.hs

```diff
@@ -0,0 +1,313 @@
+  [StoredUser] ->
+  -- | the available backend
+  Map Domain MiniBackend ->
+  -- | the down backend
+  Map Domain MiniBackend ->
+  Maybe TeamMember ->
+  UserSubsystemConfig ->
+  Sem GetUserProfileEffects a ->
+  a
+runFederationStackWithUnavailableBackends allLocalUsers availableBackends unavailableBackends teamMember cfg =
+  let unsafeError e = error $ "Unexpected error: " <> displayException e
+   in either unsafeError Imports.id
+        . runFederationStackWithUnavailableBackendsEither
+          allLocalUsers
+          availableBackends
+          unavailableBackends
+          teamMember
+          cfg
+
+runFederationStackWithUnavailableBackendsEither ::
+  [StoredUser] ->
+  -- | the available backend
+  Map Domain MiniBackend ->
+  -- | the down backend
+  Map Domain MiniBackend ->
+  Maybe TeamMember ->
+  UserSubsystemConfig ->
+  Sem GetUserProfileEffects a ->
+  Either FederationError a
+runFederationStackWithUnavailableBackendsEither allLocalUsers availableBackends unavailableBackends teamMember cfg =
+  run
+    . runError
+    . sequentiallyPerformConcurrency
+    . miniFederationAPIAccess availableBackends unavailableBackends
+    . runInputConst cfg
+    . interpretNowConst (UTCTime (ModifiedJulianDay 0) 0)
+    . evalState []
+    . inMemoryDeleteQueueInterpreter
+    . staticUserStoreInterpreter allLocalUsers
+    . miniGalleyAPIAccess teamMember
+    . runUserSubsystem cfg
+
+runNoFederationStack ::
+  [StoredUser] ->
+  Maybe TeamMember ->
+  UserSubsystemConfig ->
+  Sem GetUserProfileEffects a ->
+  a
+runNoFederationStack allUsers teamMember cfg =
+  run
+    . runErrorUnsafe
+    . sequentiallyPerformConcurrency
+    . emptyFederationAPIAcesss
+    . runInputConst cfg
+    . interpretNowConst (UTCTime (ModifiedJulianDay 0) 0)
+    . evalState []
+    . inMemoryDeleteQueueInterpreter
+    . staticUserStoreInterpreter allUsers
+    . miniGalleyAPIAccess teamMember
+    . runUserSubsystem cfg
+
+inMemoryDeleteQueueInterpreter :: Member (State [InternalNotification]) r => InterpreterFor DeleteQueue r
+inMemoryDeleteQueueInterpreter = interpret $ \case
+  EnqueueUserDeletion uid -> modify (\l -> DeleteUser uid : l)
+  EnqueueClientDeletion cid uid mConnId -> modify (\l -> DeleteClient cid uid mConnId : l)
+  EnqueueServiceDeletion pid sid -> modify (\l -> DeleteService pid sid : l)
+
+runErrorUnsafe :: Exception e => InterpreterFor (Error e) r
+runErrorUnsafe action = do
+  res <- runError action
+  case res of
+    Left e -> error $ "Unexpected error: " <> displayException e
+    Right x -> pure x
+
+emptyFederationAPIAcesss :: InterpreterFor (FederationAPIAccess MiniFederationMonad) r
+emptyFederationAPIAcesss = interpret $ \case
+  _ -> error "uninterpreted effect: FederationAPIAccess"
+
+miniFederationAPIAccess ::
+  forall a r.
+  Map Domain MiniBackend ->
+  Map Domain MiniBackend ->
+  Sem (FederationAPIAccess MiniFederationMonad : r) a ->
+  Sem r a
+miniFederationAPIAccess online _offline = do
```

Comment:
we discovered for the tests to fail on missing backends, the interpreter doesn't even have to know about the "down" backends, it will simply fail to discover the users that live on the backend with unknown domain.

Reply:
Fixed.